File: ruby/validBTS.rb
class Node
    attr_accessor :val, :left, :right
    def initialize(val = 0, left = nil, right = nil)
        @val = val
        @left = left
        @right = right
    end
end

def is_valid_bst(root)
    # type your code here
    if root.left == nil && root.right !=nil
        return is_valid_bst(root.right)
    elsif root.left != nil && root.right == nil
        return is_valid_bst(root.left)
    elsif root.left == nil && root.right == nil
        return true
    elsif root.left.val > root.val || root.right.val < root.val
        return false
    end

    boolean_array = [is_valid_bst(root.left), is_valid_bst(root.right)].filter {|boolean| boolean == false}
    if boolean_array.length == 0 
        return true
    else
        return false
    end
end

if __FILE__ == $PROGRAM_NAME

    puts "Expecting: true"
    def T(v,l,r)
        return Node.new(v, l, r)
    end
    puts "=>", is_valid_bst(T(2, T(1, nil, nil), T(3, nil, nil)))

# Don't forget to add your own!
end
  
# Please add your pseudocode to this file
# And a written explanation of your solution

File: javascript/validBTS.js
class Node {
  constructor(val, left = null, right = null){
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function isValidBST(root) {
  // type your code here
  if(root.left === null && root.right === null){
    return true
  } else if (root.left === null && root.right !== null){
    return isValidBST(root.right)
  } else if(root.left !== null && root.right === null) {
    return isValidBST(root.left)
  } else if(root.left.val > root.val || root.right.val < root.val){
    return false
  }

  const booleanArr = [isValidBST(root.left), isValidBST(root.right)].filter(boolean => boolean === false)
  return !booleanArr.length
}

if (require.main === module) {

// add your own tests in here
const T = (v, l, r) => new Node(v, l, r)

console.log("Expecting: true");
console.log("=>", isValidBST(T(2, T(1), T(3))));

}

module.exports = {
  isValidBST,
  Node
}

// Please add your pseudocode to this file
// And a written explanation of your solution